### PR TITLE
Lazily track plugin scoreboards by default

### LINKED ISSUE #4260: Scoreboard performance semi leak
###  Timings or Profile link:
**Before:** https://timings.aikar.co/?id=3e60f03df39441cd9d57d6a9b45190f9#timings
**After:** https://timings.aikar.co/?id=d9b5b0c85af34fd6b2bd2c96746b9eb1#timings

### Description of issue:
I believe I've figured out why scoreboards are causing so many performance issues. Whenever a scoreboard is created it's stored in a weak collection, however it seems to take a while for the scoreboard to be removed even when there are no references to it. Some plugins (mine at least) create many scoreboards and just rely on the scoreboard to be garbage collected. This results in basically a memory leak, but not in the sense you will run out of memory (scoreboards don't take up a lot of memory). There's just so many scoreboards in the collection it takes a very long time to search through the collection and find the score of a scoreboard. The only way I've figured out to fix this is to use reflection and remove the scoreboard from the collection after I'm done with the scoreboard. I listed the results before & after I did this change above.

I'm not really sure what you would do to fix this, maybe add a method to the API to unregister the scoreboard? Just thought I'd bring it to the attention of the developers.

### Paper version:
Observed and tested on: git-Tuinity-"ae2d7fc" (Latest 1.15.2 Tuinity) however it affects paper as well.

### PR DESCRIPTION
**Pull request by astei**
On servers with plugins that constantly churn through scoreboards, there is a risk of degraded GC performance due to the number of scoreboards held on by weak references. Most plugins don't even need the (vanilla) functionality that requires all plugin scoreboards to be tracked by the server. Instead, only track scoreboards when an objective is added with a non-dummy criteria.

This is a breaking change, however the change is a much more sensible default. In case this breaks your workflow you can always force all scoreboards to be tracked with `settings.always-track-plugin-scoreboards: true` in paper.yml.

Fixes #4260

### CODE REVIEW

File: Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch]:
```suggestion
+    private final MinecraftServer a; public final MinecraftServer getServer() { return this.a; } // Paper - OBFHELPER
```

**Review by Proximyst — DISMISSED** [review]

File: Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch
Comment (kashike) [inline Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch:81]:
```suggestion
+    void registerScoreboardForVanilla(CraftScoreboard scoreboard) {
```

**Review by kashike — DISMISSED** [review]

**Review by Proximyst — DISMISSED** [review]

File: Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch
Comment (aikar) [inline Spigot-Server-Patches/0581-Lazily-track-plugin-scoreboards-by-default.patch]: Can drop this and just use minecraftserver getserver instead, don't need a new diff when one exists to get the server already

**Review by aikar — DISMISSED** [review]
Minor nitpick, but ima mark as approved preemptively so if either you take care of it or the Dev team does it'll be clear for merge

**Comment by Proximyst** [comment]
@limbo-app ready

**Review by limbo-app — DISMISSED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by Proximyst** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Pull request merged**
